src/northbridge/amd/pi/00730F01/northbridge.c

```diff
@@ -749,6 +751,83 @@ static void northbridge_fill_ssdt_generator(const struct device *device)
 	acpigen_pop_len();
 }
 
+static void *get_drtm_tpm_log(u32 *size)
+{
+	const struct cbmem_entry *ce;
+	const u32 drtm_default_log_len = 0x1000;
+	void *lasa;
+	ce = cbmem_entry_find(CBMEM_ID_DRTM_LOG);
+	if (ce) {
+		lasa = cbmem_entry_start(ce);
+		*size = cbmem_entry_size(ce);
+		printk(BIOS_DEBUG, "DRTM log found at %p\n", lasa);
+		return lasa;
+	}
+	lasa = cbmem_add(CBMEM_ID_DRTM_LOG, drtm_default_log_len);
+	if (!lasa) {
+		printk(BIOS_ERR, "DRTM log creation failed\n");
+		return NULL;
+	}
+
+	printk(BIOS_DEBUG, "DRTM log created at %p\n", lasa);
+	memset(lasa, 0, drtm_default_log_len);
+
+	*size = drtm_default_log_len;
+	return lasa;
+}
+
+static void acpi_create_drtm_table(acpi_drtm_t *drtm)
+{
+	acpi_header_t *header = &(drtm->header);
+	unsigned long current;
+	drtm_validated_tables_list_t *drtm_vtl;
+	drtm_resources_t *drtm_resources;
+	drtm_dps_t *drtm_dps;
+
+	memset((void *)drtm, 0, sizeof(acpi_drtm_t));
+
+	/*
+	 * Some payloads like SeaBIOS depend on log area to use TPM2.
+	 * Get the memory size and address of TPM2 log area or initialize it.
+	 */
+	drtm->log_area_start = (uintptr_t)get_drtm_tpm_log(&(drtm->log_area_length));
+
+	/* Fill out header fields. */
+	memcpy(header->signature, "DRTM", 4);
+	memcpy(header->oem_id, OEM_ID, 6);
+	memcpy(header->oem_table_id, ACPI_TABLE_CREATOR, 8);
+	memcpy(header->asl_compiler_id, ASLC, 4);
+
+	header->asl_compiler_revision = asl_revision;
+	header->length = sizeof(acpi_drtm_t);
```

Comment:
Maybe set it once just before checksum as `current - header`? Right now compiler is free to mostly optimize it away, but this may change if there will be function calls in between. This would also be less prone to errors IMO.

Reply:
All tables are created like that, where additional structures are added in the meantime of creation and the size is adjusted. I would like to keep it like that.